File: irrgarten/src/irrgarten/LabyrinthCharacter.java
package irrgarten;
/**
 *
 * @author usuario
 */
public abstract class LabyrinthCharacter {
    private static final int INVALID_POS=-1;    //Variable que indica la posición inicial
    
    private String name;                        //Nombre
    private float strength;                     //Fuerza 
    private float intelligence;                 //Inteligencia
    private float health;                       //Vida actual
    private int row;                            //Fila
    private int col;                            //Columna
    
    //P5
    LabyrinthCharacter(String hownamed, float howsmart, float howstrong, float hp){
        name = hownamed;
        intelligence = howsmart;
        strength = howstrong;
        health = hp;
        row = INVALID_POS; 
        col = INVALID_POS;
    }
    
    //P5
    LabyrinthCharacter(LabyrinthCharacter other){
        name = other.name;
        intelligence = other.intelligence;
        strength = other.strength;
        health = other.health;
        row = other.row; 
        col = other.col;
    }
    
    /**
     * Funcion dead. Devuelve si está o no muerto
     * 
     * @return true si está muerto, false si no
     */
    public boolean dead(){
        return (health <= 0);
    }
    
    /**
     * Funcion getRow. Devuelve la fila en la que se encuentra
     * 
     * @return Número de fila 
     */
    public int getRow(){
         return row;
    }
    
    /**
     * Funcion getCol. Devuelve la columna en la que se encuentra 
     * 
     * @return Número de columna
     */
    public int getCol(){
        return col;
    }
    
    //P5
    protected float getIntelligence(){
        return intelligence;
    }
    
    //P5
    protected float getStrength(){
        return strength;
    }
    
    //P5
    protected float getHealth(){
        return health;
    }
    
    //P5
    protected void setHealth(float hp){
        health = hp;
    }
    
    /**
     * Función setPos. Establece la posición
     * @param r Número de fila
     * @param c Número de columna
     */
    public void setPos(int r, int c){
        row=r;
        col=c;
    }
    
    public String toString(){
        String info = name + ":\n";
        info += "\nIntelligence: " + intelligence;
        info += "\nStregth: " + strength;
        info += "\nHealth: " + health;
        info += "\nPosition: (" + row + ", " + col + ")\n\n";      
        
        return info;
    }
    
    /**
     * Funcion gotWounded. Resta uno de vida.
     */
    protected void gotWounded(){
        health--;
    }
    
    abstract public float attack();
    
    abstract public boolean defend(float attack);
}

File: Irrgarten/lib/LabyrinthCharacter.rb
class LabyrinthCharacter
	@@INVALID_POS=-1
	
	attr_reader :name 
	attr_reader :intelligence 
	attr_reader :strength 
	attr_accessor :health
	attr_reader :row
	attr_reader :col
	
	def initialize(nm, howSmart, howStrong, hp)
		@name = nm
		@intelligence = howSmart
		@strength = howStrong
		@health = hp
		@row = @@INVALID_POS
		@col = @@INVALID_POS
	end
	
	def copy(other)
		@name = other.name
		@intelligence = other.intelligence
		@strength = other.strength
		@health = other.health
		@row = other.row
		@col = other.col
	end
	
	# Devuelve true si muere el monstruo
    def dead
    	return @health <= 0;
    end
    
	# Funcion setPos. Establece la posición del jugador
    # 
    # @param r Número de fila
    # @param c Número de columna
	def setPos(r,c)
		@row=r
		@col=c
	end
	
	def to_s
		info ="\n"+@name+":\n"
		info +="\nIntelligence: " + @intelligence.to_s
		info +="\nStregth: " + @strength.to_s
		info +="\nHealth: " + @health.to_s
		info +="\nPosition: (" + @row.to_s + ", " + @col.to_s + ")"
		return info
	end
	
	def gotWounded
		@health -= 1
	end
	
	protected :intelligence, :strength, :health, :gotWounded
	private_class_method :new
end
